crates/iroha_multisig_data_model/src/lib.rs

```diff
@@ -36,39 +51,110 @@ type Weight = u8;
 /// Default multisig transaction time-to-live in milliseconds based on block timestamps
 pub const DEFAULT_MULTISIG_TTL_MS: u64 = 60 * 60 * 1_000; // 1 hour
 
-/// Arguments to propose or approve multisig transaction
-#[derive(Debug, Clone, Decode, Encode, Serialize, Deserialize, IntoSchema)]
-pub enum MultisigTransactionArgs {
-    /// Propose instructions and initialize approvals with the proposer's one
-    Propose(Vec<InstructionBox>),
-    /// Approve certain instructions
-    Approve(HashOf<Vec<InstructionBox>>),
+/// Propose a multisig transaction and initialize approvals with the proposer's one
+#[derive(Debug, Clone, Encode, Decode, Serialize, Deserialize, IntoSchema, Constructor)]
+pub struct MultisigPropose {
+    /// Multisig account to propose
+    pub account: AccountId,
+    /// Proposal contents
+    pub instructions: Vec<InstructionBox>,
 }
 
-impl From<MultisigAccountArgs> for Json {
-    fn from(details: MultisigAccountArgs) -> Self {
-        Json::new(details)
-    }
+/// Approve a certain multisig transaction
+#[derive(Debug, Clone, Encode, Decode, Serialize, Deserialize, IntoSchema, Constructor)]
+pub struct MultisigApprove {
+    /// Multisig account to approve
+    pub account: AccountId,
+    /// Proposal to approve
+    pub instructions_hash: HashOf<Vec<InstructionBox>>,
 }
 
-impl TryFrom<&Json> for MultisigAccountArgs {
-    type Error = serde_json::Error;
+macro_rules! impl_custom_instruction {
```

Comment:
I believe we could introduce some `#[derive(Instruction)]` macro for custom ISI. We usually provide a derive macro for custom stuff. Just a though, not for this PR

Reply:
Yes, this can be considered as a previous step to a derive macro